corporate: Send a reminder email 2 months before the end date.

prakhar1144 opened the pull request:
* Commit 1: Prep commit -- explanation in commit message.
* Commit 2: Set `end_date` for fixed-price plans.
* Commit 3: Send a reminder email to sales@zulip.com 2 months before the end_date.

We'll have to do a [db migration](https://github.com/zulip/zulip/pull/28934#issuecomment-1953899293) to set the `end_date` of existing fixed-price plans. 

<details><summary>Email</summary>
<img src="https://github.com/zulip/zulip/assets/56781761/08caf16e-af95-43bb-8c1d-0bb8978f0bc9">
</img>
</details> 

<details>
<summary>Self-review checklist</summary>

<!-- Prior to submitting a PR, follow our step-by-step guide to review your own code:
https://zulip.readthedocs.io/en/latest/contributing/code-reviewing.html#how-to-review-code -->

<!-- Once you create the PR, check off all the steps below that you have completed.
If any of these steps are not relevant or you have not completed, leave them unchecked.-->

- [x] [Self-reviewed](https://zulip.readthedocs.io/en/latest/contributing/code-reviewing.html#how-to-review-code) the changes for clarity and maintainability
      (variable names, code reuse, readability, etc.).

Communicate decisions, questions, and potential concerns.

- [x] Explains differences from previous plans (e.g., issue description).
- [x] Highlights technical choices and bugs encountered.
- [x] Calls out remaining decisions and concerns.
- [x] Automated tests verify logic where appropriate.

Individual commits are ready for review (see [commit discipline](https://zulip.readthedocs.io/en/latest/contributing/commit-discipline.html)).

- [x] Each commit is a coherent idea.
- [x] Commit message(s) explain reasoning and motivation for changes.

Completed manual review and testing of the following:

- [ ] Visual appearance of the changes.
- [ ] Responsiveness and internationalization.
- [ ] Strings and tooltips.
- [ ] End-to-end functionality of buttons, interactions and flows.
- [ ] Corner cases, error conditions, and easily imagined bugs.
</details>

prakhar1144 commented on corporate/models.py:
```diff
@@ -353,6 +353,11 @@ class CustomerPlan(AbstractCustomerPlan):
+    # Flag to track if an email has been sent to Zulip team to
+    # review the pricing, 60 days before the end date. Helps to send
+    # an email only once and not every time when cron run.
+    reminder_to_review_plan_email_sent = models.BooleanField(default=False)
```
While naming this field, I didn't included the term "fixed-price-plan" because, I think, we can reuse this field to track reminder email sent for Enterprise plans.

timabbott replied:
Yeah, it seems like a reasonable generic mechanism.

timabbott commented on corporate/lib/stripe.py:
```diff
@@ -1920,11 +1929,7 @@ def make_end_of_cycle_updates_if_needed(
-                and plan.end_date == next_billing_cycle
-                and plan.status == CustomerPlan.ACTIVE
-            ):
+            if plan.end_date == next_billing_cycle and plan.status == CustomerPlan.ACTIVE:
```
@amanagr remind me why we had this logic conditional on the legacy plan tier previously?

I also wonder if this logic is correct in the case that there's a next plan, or if it could result in a problematic intermediate downgrade that isn't intended.

prakhar1144 replied:
If there's a next plan, the status of current_plan would be `SWITCH_PLAN_TIER_AT_PLAN_END`

amanagr replied:
Only reason I had that check because we didn't had plans to support anything like this at the time.

timabbott commented on corporate/lib/stripe.py:
```diff
@@ -1667,6 +1675,7 @@ def process_initial_upgrade(
                 # Manual license management is not available for fixed price plan.
                 assert automanage_licenses is True
                 plan_params["fixed_price"] = fixed_price_plan_offer.fixed_price
+                plan_params["end_date"] = add_months(billing_cycle_anchor, 12)
```
Should we make this 12 be a `CustomerPlanOffer.BILLING_CYCLE_DURATION_MONTHS` that's hardcoded to 12 -- just to help clarify in the plumbing that this might in the future be a parameter? I'm not sure we can do the same for the other 12, but this seems right.

prakhar1144 replied:
Added in `CustomerPlan`:
```
    # Currently, all the fixed-price plans are configured for one year.
    # In future, we might change this to a field.
    FIXED_PRICE_PLAN_DURATION_MONTHS = 12
```

timabbott commented on corporate/lib/stripe.py:
```diff
@@ -4768,6 +4768,25 @@ def invoice_plans_as_needed(event_time: Optional[datetime] = None) -> None:
+                    context=context,
+                )
+                plan.reminder_to_review_plan_email_sent = True
+                plan.save(update_fields=["reminder_to_review_plan_email_sent"])
```
We probably need code to set this flag back to `False` through whatever mechanism we expect end dates to be mutated/extended. (Presumably, if they move onto a new plan at the end date, it's fine, but if we extend the end date another yet, we wouldn't find out)

prakhar1144 replied:
Set it to `False` in `update_end_date_of_current_plan`

prakhar1144 commented:
Updates in the recent push:
* Addressed the review comments -- see respective comments
* Added a prep commit (first commit) -- explanation in commit message.
* Added a comment in the last commit to be explicit about what `62` means:
```diff
elif (
    plan.fixed_price is not None
    and not plan.reminder_to_review_plan_email_sent
    and plan.end_date is not None  # for mypy
+   # The max gap between two months is 62 days. (1 Jul - 1 Sep)
    and plan.end_date - plan.next_invoice_date <= timedelta(days=62)
):
```
* Migration to run on cloud:
```python
from corporate.models import CustomerPlan
from corporate.lib.stripe import add_months

fixed_price_plans = CustomerPlan.objects.filter(status=CustomerPlan.ACTIVE, fixed_price__isnull=False)
for plan in fixed_price_plans:
    assert plan.end_date is None
    assert plan.billing_cycle_anchor is not None
    plan.end_date = add_months(plan.billing_cycle_anchor, 12)
    plan.save(update_fields=["end_date"])
```

timabbott commented on corporate/lib/stripe.py:
```diff
@@ -4832,6 +4858,26 @@ def invoice_plans_as_needed(event_time: Optional[datetime] = None) -> None:
                     plan.invoice_overdue_email_sent = True
                     plan.save(update_fields=["invoice_overdue_email_sent"])
                 continue
```
Hmm, I think this `continue/elif` construction creates risk that if a server isn't sending audit log updates, which seems possible for some cases of a fixed-price plan, then we might not run the new `elif` block.

One theory is that the new `elif` block should just be an independent `if` block placed before this one.

prakhar1144 replied:
That's right. 

> the new elif block should just be an independent if block placed before this one.

Made the changes.

prakhar1144 commented:
Updates in the recent push:
* Addressed the review comment.
* Added a migration file [as discussed](https://github.com/zulip/zulip/pull/28934#issuecomment-1954853995) -- See 2nd commit.

timabbott commented:
Merged, since I'd like to be able to stack this with other work, but please consider https://github.com/zulip/zulip/pull/28934#discussion_r1498113277 as a potential follow-up.
